aexpect/client.py

```diff
@@ -277,6 +277,28 @@ def _close_reader_fds(self):
             except OSError:
                 pass
 
+    def _close_aexpect_helper(self):
+        """
+        Makes sure the subprocess holding the aexpect_helper is terminated
+        and that the file descriptors related to it are closed. Then releases
+        the popen instance.
+        """
+        if self._aexpect_helper:
+            try:
+                try:
+                    self._aexpect_helper.wait(10)
+                except subprocess.TimeoutExpired:
+                    self._aexpect_helper.kill()
+            finally:
+                if self._aexpect_helper.stdout:
+                    self._aexpect_helper.stdout.close()
+                try:
+                    if self._aexpect_helper.stdin:
+                        self._aexpect_helper.stdin.close()
+                finally:
+                    self._aexpect_helper.wait(10)
+                self._aexpect_helper = None
```

Comment:
This cleanup looks a bit unnecessary complicated to me, how about simply:

```python
if self._aexpect_helper:
    try:
        self._aexpect_helper.wait(10)
    except subprocess.TimeoutExpired:
        self._aexpect_helper.kill()
    finally:
        if self._aexpect_helper.stdout:
            self._aexpect_helper.stdout.close()
        if self._aexpect_helper.stdin:
            self._aexpect_helper.stdin.close()
        self._aexpect_helper.wait(10)
    self._aexpect_helper = None
```

I mean the second try/finally only catches stdin handling and not the stdout handling.

Reply:
Hi,

As I commented before, considering your feedback and getting code simplified, even something like 
```python
    def _close_aexpect_helper(self):
        """
        Makes sure the subprocess holding the aexpect_helper is terminated
        and that the file descriptors related to it are closed. Then releases
        the popen instance.
        """
        if self._aexpect_helper:
            try:
                self._aexpect_helper.wait(10)
            except subprocess.TimeoutExpired:
                self._aexpect_helper.kill()
                self._aexpect_helper.wait(10)
            finally:
                self._aexpect_helper = None
```

Should have the work done by itself, as the problem was that saving an indirect reference to the process wasn't letting the python garbage collector clean the `Popen` related to the `self._aexpect_helper`. However, this way we wouldn't be manually assuring that the fds are closed, but relying on the Popen's cleanup routines instead.

According to some testing I'm running right now, even the now suggested code piece is closing the opened fds. Deciding whether one or the other fits better on the code should be related to wanting to get things done _manually_ or just relying on the `Popen` cleanup routines implemented on `subprocess.py`.

Please, @ldoktor, @luckyh let me know what do you think about this. I'll update the commit considering the feedback.

Thank you,
Beñat

Reply:
@bgartzi yes, I had realized this thing when I was going to give the suggestion https://github.com/avocado-framework/aexpect/pull/97#discussion_r783689453. But finally, for making the behavior predictable, I decided to not let the cleanup procedure rely on Python's GC. Or otherwise, I'm not sure if we might deal with some weird problem related to that in the future.

Thanks,